velox/exec/fuzzer/MemoryArbitrationFuzzer.cpp

```diff
@@ -0,0 +1,448 @@
+      core::JoinType::kFull,
+      core::JoinType::kLeftSemiFilter,
+      core::JoinType::kLeftSemiProject,
+      core::JoinType::kAnti};
+  std::vector<const core::PlanNodePtr> plans;
+  plans.reserve(kJoinTypes.size());
+  for (const auto& joinType : kJoinTypes) {
+    plans.push_back(hashJoinPlan(joinType));
+  }
+  return plans;
+}
+
+std::vector<core::PlanNodePtr> MemoryArbitrationFuzzer::aggregatePlans() {
+  const auto numKeys = randInt(1, 5);
+  // Reuse the hash join utilites to generate aggregation keys and inputs.
+  const std::vector<TypePtr> keyTypes = generateJoinKeyTypes(numKeys);
+  std::vector<std::string> groupingKeys = makeNames("g", keyTypes.size());
+  const auto aggregateInputs = generateProbeInput(groupingKeys, keyTypes);
+  const std::vector<std::string> aggregates{"count(1)"};
+  std::vector<core::PlanNodePtr> plans;
+  plans.push_back(PlanBuilder()
+                      .values(aggregateInputs)
+                      .singleAggregation(groupingKeys, aggregates, {})
+                      .planNode());
+  // Partial -> final aggregation plan.
+  plans.push_back(PlanBuilder()
+                      .values(aggregateInputs)
+                      .partialAggregation(groupingKeys, aggregates, {})
+                      .finalAggregation()
+                      .planNode());
+  // Partial -> intermediate -> final aggregation plan.
+  plans.push_back(PlanBuilder()
+                      .values(aggregateInputs)
+                      .partialAggregation(groupingKeys, aggregates, {})
+                      .intermediateAggregation()
+                      .finalAggregation()
+                      .planNode());
+  return plans;
+}
+
+void MemoryArbitrationFuzzer::verify() {
+  const auto outputDirectory = TempDirectoryPath::create();
+  const auto spillDirectory = exec::test::TempDirectoryPath::create();
+  std::vector<core::PlanNodePtr> plans;
+
+  for (const auto& plan : hashJoinPlans()) {
+    plans.push_back(plan);
+  }
+
+  for (const auto& plan : aggregatePlans()) {
+    plans.push_back(plan);
+  }
+
+  SCOPE_EXIT {
+    waitForAllTasksToBeDeleted();
+  };
+
+  int successCount = 0;
+  int failCount = 0;
+  const auto numThreads = plans.size() * 8;
+  std::vector<std::thread> queryThreads;
+  queryThreads.reserve(numThreads);
+  for (int i = 0; i < numThreads; ++i) {
+    queryThreads.emplace_back([&, i]() {
+      try {
+        const auto memoryCapacity = static_cast<int64_t>(randInt(1, 32)) * 4L;
+        auto queryCtx = newQueryCtx(
+            memory::MemoryManager::getInstance(),
+            executor_.get(),
+            memoryCapacity << 20);
+        auto queryConfigs = queryConfigsWithSpill_;
+        if (vectorFuzzer_.coinToss(0.1)) {
+          queryConfigs = {{core::QueryConfig::kSpillEnabled, "false"}};
+        }
+
+        AssertQueryBuilder(plans.at(i % plans.size()))
+            .configs(queryConfigs)
+            .spillDirectory(spillDirectory->getPath() + fmt::format("/{}/", i))
```

Comment:
I think the spill is disabled without spillDirectory. So we can just have one queryConfigWithSpill_